src/flint/types/nmod.pyx

```diff
@@ -99,100 +255,108 @@ cdef class nmod(flint_scalar):
+            r.ctx = s2.ctx
+            r.val = nmod_add(val, s2.val, s2.ctx.mod)
             return r
         return NotImplemented
 
     def __radd__(s, t):
-        cdef nmod r
+        cdef nmod r, s2
         cdef mp_limb_t val
-        if any_as_nmod(&val, t, (<nmod>s).mod):
+        s2 = s
+        if s2.ctx.any_as_nmod(&val, t):
             r = nmod.__new__(nmod)
-            r.mod = (<nmod>s).mod
-            r.val = nmod_add((<nmod>s).val, val, r.mod)
+            r.ctx = s2.ctx
+            r.val = nmod_add(s2.val, val, s2.ctx.mod)
             return r
         return NotImplemented
 
     def __sub__(s, t):
-        cdef nmod r
+        cdef nmod r, s2
         cdef mp_limb_t val
-        if any_as_nmod(&val, t, (<nmod>s).mod):
+        s2 = s
+        if s2.ctx.any_as_nmod(&val, t):
             r = nmod.__new__(nmod)
-            r.mod = (<nmod>s).mod
-            r.val = nmod_sub((<nmod>s).val, val, r.mod)
+            r.ctx = s2.ctx
+            r.val = nmod_sub(s2.val, val, s2.ctx.mod)
             return r
         return NotImplemented
 
     def __rsub__(s, t):
         cdef nmod r
         cdef mp_limb_t val
-        if any_as_nmod(&val, t, (<nmod>s).mod):
+        s2 = s
+        if s2.ctx.any_as_nmod(&val, t):
             r = nmod.__new__(nmod)
-            r.mod = (<nmod>s).mod
-            r.val = nmod_sub(val, (<nmod>s).val, r.mod)
+            r.ctx = s2.ctx
+            r.val = nmod_sub(val, s2.val, s2.ctx.mod)
             return r
         return NotImplemented
 
     def __mul__(s, t):
-        cdef nmod r
+        cdef nmod r, s2
         cdef mp_limb_t val
-        if any_as_nmod(&val, t, (<nmod>s).mod):
+        s2 = s
+        if any_as_nmod(&val, t, s2.ctx.mod):
             r = nmod.__new__(nmod)
-            r.mod = (<nmod>s).mod
-            r.val = nmod_mul(val, (<nmod>s).val, r.mod)
+            r.ctx = s2.ctx
+            r.val = nmod_mul(val, s2.val, s2.ctx.mod)
             return r
         return NotImplemented
 
     def __rmul__(s, t):
-        cdef nmod r
+        cdef nmod r, s2
         cdef mp_limb_t val
-        if any_as_nmod(&val, t, (<nmod>s).mod):
+        s2 = s
+        if s2.ctx.any_as_nmod(&val, t):
             r = nmod.__new__(nmod)
-            r.mod = (<nmod>s).mod
-            r.val = nmod_mul((<nmod>s).val, val, r.mod)
+            r.ctx = s2.ctx
+            r.val = nmod_mul(s2.val, val, s2.ctx.mod)
             return r
         return NotImplemented
 
     @staticmethod
     def _div_(s, t):
-        cdef nmod r
+        cdef nmod r, s2, t2
         cdef mp_limb_t sval, tval, x
-        cdef nmod_t mod
+        cdef nmod_ctx ctx
         cdef ulong tinvval
 
         if typecheck(s, nmod):
-            mod = (<nmod>s).mod
-            sval = (<nmod>s).val
-            if not any_as_nmod(&tval, t, mod):
+            s2 = s
+            ctx = s2.ctx
+            sval = s2.val
+            if not ctx.any_as_nmod(&tval, t):
                 return NotImplemented
```

Comment:
Calling `ctx.any_as_nmod` method is slower than calling the `any_as_nmod` function...

Reply:
If we declare them as `@cython.final` to avoid the virtual method table then it is just as fast I think...